File: src/Python/101-200/161.StringOneEditDistance.py

class Solution:
    def isOneEditDistance(self, s, t):
        if s is None and t is None:
            return False
        if s is None:
            return len(t) - 1
        if t is None:
            return len(s) - 1
        if len(s) < len(t):
            s, t = t, s
        if len(s) == 1 and len(t) == 0:
            return True
        if len(s) > len(t) + 1:
            return False
        if len(s) == len(t):
            dist = 0
            for i in range(0, len(s)):
                if s[i] != t[t]:
                    dist += 1
                    if dist > 1:
                        return False
        else:
            dist = 0 
            for i in range(0, len(t)):
                if t[i] == s[i + dist]:
                    continue
                dist += 1
                if dist > 1:
                    return False
                if t[i] == s[i + dist]:
                    continue
                return dist <= 1
        


if __name__ == "__main__":
    solution = Solution()
    print(solution.isOneEditDistance("ss", "sst"))

File: src/Java/101-200/161.StringOneEditDistance.java

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;



public class Solution {
    public boolean isOneEditDistance(String s, String t) {
        if (s == null && t == null) return false;
        if (s == null) return t.length() == 1;
        if (t == null) return s.length() == 1;
        if (s.length() < t.length()) {
            String temp = s;
            s = t;
            t = temp;
        }
        if (s.length() == 1 && t.length() == 0) return true;
        if (s.length() > t.length() + 1) return false;
        if (s.length() == t.length()) {
            int dist = 0;
            for(int i=0; i<s.length(); i++) {
                if (s.charAt(i) != t.charAt(i)) {
                    dist ++;
                    if (dist > 1) return false;
                }
            }
            return dist == 1;
        } else {
            int dist = 0;
            for(int i=0; i<t.length(); i++) {
                if (t.charAt(i) == s.charAt(i+dist)) continue;
                dist ++;
                if (dist > 1) return false;
                if (t.charAt(i) == s.charAt(i+dist)) continue;
                return false;
            }
            return dist <= 1;
        }
    }
}


